integration/integration.go

```diff
@@ -0,0 +1,244 @@
+	tchannelClusterAddr    = flag.String("clustertchanneladdr", "0.0.0.0:9001", "Cluster TChannel server address")
+	httpNodeAddr           = flag.String("nodehttpaddr", "0.0.0.0:9002", "Node HTTP server address")
+	tchannelNodeAddr       = flag.String("nodetchanneladdr", "0.0.0.0:9003", "Node TChannel server address")
+	errServerStartTimedOut = errors.New("server took too long to start")
+	errServerStopTimedOut  = errors.New("server took too long to stop")
+)
+
+type dataMap map[string][]m3db.Datapoint
+
+func databaseOptions() m3db.DatabaseOptions {
+	var opts m3db.DatabaseOptions
+	opts = storage.NewDatabaseOptions().NewBootstrapFn(func() m3db.Bootstrap {
+		return bootstrap.NewNoOpBootstrapProcess(opts)
+	})
+	return opts
+}
+
+func setup() (m3db.DatabaseOptions, *time.Time, error) {
+	dbOpts := databaseOptions()
+
+	// set up now fn
+	now := time.Now().Truncate(dbOpts.GetBlockSize())
+	nowFn := func() time.Time { return now }
+	dbOpts = dbOpts.NowFn(nowFn)
+
+	// set up file path
+	dir, err := ioutil.TempDir("", "integration-test")
+	if err != nil {
+		return nil, nil, err
+	}
+	dbOpts = dbOpts.FilePathPrefix(dir)
+
+	return dbOpts, &now, nil
+}
+
+type conditionFn func() bool
+
+func waitUntil(fn conditionFn, timeout time.Duration) bool {
+	deadline := time.Now().Add(timeout)
+	for time.Now().Before(deadline) {
+		if fn() {
+			return true
+		}
+		time.Sleep(time.Second)
+	}
+	return false
+}
+
+func waitUntilServerIsUp(address string) error {
+	fakeRequest := rpc.NewFetchRequest()
+	serverIsUp := func() bool { _, err := fetch(fakeRequest); return err == nil }
+	if waitUntil(serverIsUp, serverStateChangeTimeout) {
+		return nil
+	}
+	return errServerStartTimedOut
+}
+
+func waitUntilServerIsDown(address string) error {
+	fakeRequest := rpc.NewFetchRequest()
+	serverIsDown := func() bool { _, err := fetch(fakeRequest); return err != nil }
+	if waitUntil(serverIsDown, serverStateChangeTimeout) {
+		return nil
+	}
+	return errServerStopTimedOut
+}
+
+func startServer(opts m3db.DatabaseOptions, doneCh chan struct{}) error {
+	go server.Serve(*httpClusterAddr, *tchannelClusterAddr, *httpNodeAddr, *tchannelNodeAddr, opts, doneCh)
+
+	return waitUntilServerIsUp(*tchannelNodeAddr)
+}
+
+func stopServer(doneCh chan<- struct{}) error {
+	doneCh <- struct{}{}
+
+	return waitUntilServerIsDown(*tchannelNodeAddr)
+}
+
+func generateTestData(metricNames []string, numPoints int, start time.Time) dataMap {
+	if numPoints <= 0 {
+		return nil
+	}
+	testData := make(map[string][]m3db.Datapoint)
+	for _, name := range metricNames {
+		datapoints := make([]m3db.Datapoint, 0, numPoints)
+		for i := 0; i < numPoints; i++ {
+			timestamp := start.Add(time.Duration(i) * time.Second)
+			datapoints = append(datapoints, m3db.Datapoint{
+				Timestamp: timestamp,
+				Value:     0.1 * float64(i),
+			})
+		}
+		testData[name] = datapoints
+	}
+	return testData
+}
+
+func tchannelClient(address string) (xclose.SimpleCloser, rpc.TChanNode, error) {
```

Comment:
FYI I ended up using an option to control whether we use the tchannel client or the m3db client to read/write data. Feels like that gives us more flexibility in terms of testing things.

For now I'm using the m3db client for writing and the tchannel client for reading by default. Once the client read diff lands, I can integrate it into the testing setup and switch it over to using the m3db client for reading by default as well.